### PR TITLE
feat: error tracking sparklines

### PR DESCRIPTION
**Pull request by daibhin**
## Problem

Who doesn't love a sparkline

## Changes

Screenshot 2024-06-21 at 17 40 53

TODO:
- Can we allow `arrayReduce`?
- Is there a better way?

### CODE REVIEW

File: posthog/hogql/functions/mapping.py [hunk 1]
```diff
@@ -325,7 +325,9 @@ def compare_types(arg_types: list[ConstantType], sig_arg_types: tuple[ConstantTy
     "arrayDistinct": HogQLFunctionMeta("arrayDistinct", 1, 1),
     "arrayEnumerateDense": HogQLFunctionMeta("arrayEnumerateDense", 1, 1),
     "arrayIntersect": HogQLFunctionMeta("arrayIntersect", 1, None),
-    # "arrayReduce": HogQLFunctionMeta("arrayReduce", 2,None),  # takes a "parametric function" as first arg, is that safe?
```
Comment (daibhin) [inline posthog/hogql/functions/mapping.py:328]: @Twixes looks like this comment was added by you. Do you know if it's safe to add the `arrayReduce` function yet?

File: frontend/src/scenes/error-tracking/ErrorTrackingScene.tsx [hunk 2]
```diff
@@ -32,6 +33,10 @@ export function ErrorTrackingScene(): JSX.Element {
                 width: '50%',
                 render: CustomGroupTitleColumn,
             },
+            "toJSONString(arrayReduce('sumMap', arrayMap(x -> map(x,1), groupArray(toHour(timestamp))))) -- Sparkline":
```
Comment (daibhin) [inline frontend/src/scenes/error-tracking/ErrorTrackingScene.tsx:36]: @pauldambra mainly created this to get feedback from you on the best way to generate this data in a ClickHouse query.

Basically here:
- `groupArray` gives us an array of every hour an event happened in e.g. `[0,23,16,12,12,12,12,13,13]`
- Uses `arrayMap` to create a map of counts from each value  e.g. `[{0: 1},{23: 1},{16: 1}, {12: 1}{12: 1}{12: 1}{12: 1}{13: 1}{13: 1}]`
- `arrayReduce('sumMap', ...)` can then combine all those maps together to give e.g. `{0: 1, 23:1, 16:1, 12:3, 13: 2}`

There's still work to be done on the frontend then because the ideal structure is an array of counts zero padded for missing hours (done in `SparklineColumn`).

Perhaps I'm vastly overcomplicating this but I'm maybe not familiar enough with the ClickHouse functions to know better

Comment (daibhin) [reply frontend/src/scenes/error-tracking/ErrorTrackingScene.tsx:36]: The other option is to return the results of the `groupArray` but ideally we'd aggregate in CH because that array could be quite large for very frequent exceptions

Comment (pauldambra) [reply frontend/src/scenes/error-tracking/ErrorTrackingScene.tsx:36]: big +1 to doing as much as possible in CH
particularly if you end up with large datasets it's soooo much cheaper than looping and summing in python

File: posthog/hogql/functions/mapping.py (hunk 1 above)
Comment (Twixes) [reply posthog/hogql/functions/mapping.py:328]: I _think_ yes, but it was actually @mariusandra in [here](https://github.com/PostHog/posthog/commit/692fa66e6390af9bb45c38213b913e4ef22f729e#diff-5f19b03cbc6b7f294cabe74afca9b139317c28f3534f35bf8c1bac351ef6ed26R58-R59), so I'll defer to him

File: frontend/src/scenes/error-tracking/ErrorTrackingScene.tsx (hunk 2 above)
Comment (Twixes) [reply frontend/src/scenes/error-tracking/ErrorTrackingScene.tsx:36]: Fly-by note: I would call this column not for the visualization type ("Sparkline"), but for the data represented (e.g. "Volume by hour")

File: posthog/hogql/functions/mapping.py (hunk 1 above)
Comment (mariusandra) [reply posthog/hogql/functions/mapping.py:328]: Oh dear... hmm... those might be safe, but we don't know. Somebody needs to dig into the ClickHouse C++ code and verify. I haven't had the time to do that yet.

My biggest worry is of course that users will be able to call things that they shouldn't, and such "magic function name constants" share the 💩 out of me: `arrayReduce('sumMap', ...)`. Other than `sumMap`, can we be 100% certain that anything users pass in there is safe? I can't.

If we enable this function, we'll at the very least need to be extremely prudent about what strings we allow as the first argument of the function, and ideally even make sure that the arguments passed match a certain format (min/max length?).

Comment (mariusandra) [reply posthog/hogql/functions/mapping.py:328]: I think the easiest solution is to enable `arrayReduce`, but limit the first argument to a list of allowed constants. It's not as hard to implement as you might think: all we need is [some logic in printer.py](https://github.com/PostHog/posthog/blob/ec59e6f43d6a5550548db1cf7ce2b1e018a3e854/posthog/hogql/printer.py#L716) that makes sure that for it, the first argument must equals `ast.Constant(value='sumMap')` or any of the other allowed constants...

Comment (daibhin) [reply posthog/hogql/functions/mapping.py:328]: Prepped this in https://github.com/PostHog/posthog/pull/23261 (it really wasn't that difficult)

Comment (daibhin) [reply posthog/hogql/functions/mapping.py]: Resolved in https://github.com/PostHog/posthog/commit/7d7784251adbd3077cfbb66826de2d6e6cd95f7f#r143578812

File: frontend/src/scenes/error-tracking/ErrorTrackingScene.tsx (hunk 2 above)
Comment (daibhin) [reply frontend/src/scenes/error-tracking/ErrorTrackingScene.tsx]: Ended up with a nicer implementation in https://github.com/PostHog/posthog/commit/7d7784251adbd3077cfbb66826de2d6e6cd95f7f

Good call on the naming... updated

File: frontend/src/queries/nodes/DataTable/renderColumn.tsx [hunk 3]
```diff
@@ -79,26 +102,6 @@ export function renderColumn(
         }
         if (typeof value === 'object') {
             if (Array.isArray(value)) {
-                if (value[0] === '__hogql_chart_type' && value[1] === 'sparkline') {
```
Comment (daibhin) [inline frontend/src/queries/nodes/DataTable/renderColumn.tsx:82]: Not contained within a HogQL query in this case

File: frontend/src/scenes/error-tracking/queries.ts [hunk 4]
```diff
@@ -21,6 +21,7 @@ export const errorTrackingQuery = ({
             select: [
                 'any(properties) -- Error',
                 'properties.$exception_type',
+                'sparkline(arrayMap(number -> countEqual(groupArray(toHour(timestamp)), number), range(24))) -- Volume',
```
Comment (pauldambra) [inline frontend/src/scenes/error-tracking/queries.ts:24]: can you ELI5?
the range(24) makes me wonder what this would do at different time of the day (but I can't wrap my head around what it's doing right now so asking rather than figuring out :))

**Review by pauldambra — APPROVED** [review]
👍

File: frontend/src/scenes/error-tracking/queries.ts [hunk 5]
```diff
@@ -21,6 +21,7 @@ export const errorTrackingQuery = ({
             select: [
                 'any(properties) -- Error',
                 'properties.$exception_type',
+                'sparkline(reverse(arrayMap(x -> countEqual(groupArray(dateDiff("hour", now() - INTERVAL 1 day, timestamp)), x), range(24)))) -- Volume',
```
Comment (daibhin) [inline frontend/src/scenes/error-tracking/queries.ts:24]: @pauldambra I ended up rewriting the query because you made a great point in https://github.com/PostHog/posthog/pull/23160/commits/e128cce253a348dbc8af24e03cfc3f8ec2389e59#r1655641858...

`toHour(timestamp)` gave the hour the event happened which might incorrectly attribute values for events that happen in the same hour but on different days e.g. if the date range is 2 days and two events happen at 6pm but on different days that would end up being `[6,6]` which would get grouped together as `[0, 0, 0, 0, 0, 2, 0, 0, ... (0's up to the 24th index)]` (2 represents the count of the errors that happened in the hour at that index).

---

The right thing to do is a relative date match:
```
dateDiff('hour', timestamp, now())
```
This will give us the number of hours since now() that the event occurred.

The `groupArray` makes an array of the values for every event in the selection, so you end up with something like `[0,0,1,2,3,4,24,3,48,2,70]` which represent the number of hours since each individual exception occurred.

The `range(24)` can be thought of as an empty array with a single value for every hour in the day. For the first version I'm only going to show results from the last day grouped by hour (hence the 24) but we might want to extend that in time.

We use `arrayMap` to loop over the 24 items in the array. `countEqual` finds the number of events that happened in that hour of the 24 by summing the occurrences of that hour (variable `x`) in the grouped array.

Because we want the more recent events to appear later (e.g. events that happened 0 hours ago at the end of the sparkline) we need to reverse the result.

`sparkline` is then just a visual wrapper function to display it nicely in the data table on the frontend

Comment (pauldambra) [reply frontend/src/scenes/error-tracking/queries.ts:24]: awesome explanation, thank you!
